### ISSUE TITLE
4 vm creation - all get the same id and it fails. new version "3.0.1-rc4"

### ISSUE THREAD

**Issue by zuka1337** [Telmate/terraform-provider-proxmox#1136]
Tring to create 4 machines.
Terraform apply try to create all 4 with the same id and it fails:

![image](https://github.com/user-attachments/assets/56aed8e0-4b13-4f69-b571-b9963b3403be)
![image](https://github.com/user-attachments/assets/8c100686-4def-46fe-9351-c87e2dc11712)
![image](https://github.com/user-attachments/assets/5561a460-b481-4564-a70b-fdcf730fec14)

Simple example, there is also a second part of the "loop" to create other two machines.
```
resource "proxmox_vm_qemu" "cloudinit-k3s-master" {
    # Node name has to be the same name as within the cluster
    # this might not include the FQDN
    target_node = "home"
    desc = "Cloudinit Ubuntu"
    count = 2
    onboot = true

    # The template name to clone this vm from
    clone = "ubuntu-cloud"

    # Activate QEMU agent for this VM
    agent = 0

    os_type = "cloud-init"
    cores = 2
    sockets = 2
    numa = true
    vcpus = 0
    cpu = "host"
    memory = 2048
    name = "k3s-master-0${count.index + 1}"

    #cloudinit_cdrom_storage = "local-lvm"
    scsihw   = "virtio-scsi-pci" 
    bootdisk = "scsi0"
    boot = "order=scsi0;ide2"


    disks {
        ide {
            ide2 {
                cloudinit {
                    storage = "local-lvm"
                    
                }
            }
        }
        scsi {
            scsi0 {
                disk {
                    size            = "22732M"
                    #cache           = "writeback"
                    storage         = "local-lvm"
                    #storage_type    = "rbd"
                    #iothread        = true
                    #discard         = true
                }
            }
        }
    }

    # Setup the ip address using cloud-init.
    # Keep in mind to use the CIDR notation for the ip.
    ipconfig0 = "ip=192.x${count.index + 1}/24,gw=192.x"
    ciuser = "var.user"
    nameserver = "192.x"
    sshkeys = var.ssh
}
```

**Comment by zuka1337** [comment]
Adding   vmid field solve the issue, but is not practical if we use multiple terraform projects.
It is nonsense to map vmid each time you create a new vm:
![image](https://github.com/user-attachments/assets/3c71f13a-1ac4-4bd8-8d1e-8bee08487ca0)

**Comment by hayer** [comment]
Got into the same issue, currently running with 
```
provider "proxmox" {
    pm_parallel = 1
    # ...
}
```

**Comment by Tinyblargon** [comment]
@zuka1337 
for now the solution is what @hayer suggests.
The treading at the core of the provider needs to be reworked.

Probably best for now to make `pm_parrallel` default to `1`, and put a warning in the description that setting it higher causes issues in certain scenarios.

**Comment by TomyLobo** [comment]
Here's what I got using an MITM proxy:
|ID|Source|Req. Timestamp|Method|URL|Code|Reason|RTT|Size Resp. Body|Highest Alert|Note|Tags|
|--|------|--------------|------|---|----|------|---|---------------|-------------|----|----|
|89|Proxy|Thu Nov 28 19:51:43 CET 2024|GET|https://proxmox-01.company.tld:8006/api2/json/cluster/nextid|200|OK|18|14||false|
|92|Proxy|Thu Nov 28 19:51:44 CET 2024|GET|https://proxmox-01.company.tld:8006/api2/json/cluster/nextid|200|OK|18|14|Low|false|JSON
|93|Proxy|Thu Nov 28 19:51:44 CET 2024|POST|https://proxmox-01.company.tld:8006/api2/json/nodes/proxmox-01/qemu|200|OK|50|111||false|
|96|Proxy|Thu Nov 28 19:51:44 CET 2024|POST|https://proxmox-01.company.tld:8006/api2/json/nodes/proxmox-01/qemu|500|unable to create VM 105 - VM 105 already exists on node 'proxmox-01'|1101|13|Low|false|JSON

Probably just needs a mutex of some sort there, to ensure that there isn't another nextid invocation between the first nextid and the associated VM creation.
Probably not telling you anything new :)

**Comment by Tinyblargon** [comment]
@TomyLobo thanks for your suggestion. I'm gonna look into what needs to change so we can do this.

**Comment by zapotocnylubos** [comment]
Hi, I am testing `3.0.1-rc6` and I believe, that we encountered the same problem in our setup today.

We created about 40 VMs in 5 minutes (we use [Coder](https://github.com/coder/coder) with this Proxmox provider) for student "playground" environments.

During the machine's spin-up, two students somehow got mismatched IDs returned to the Coder/Terraform initialization. The Coder then thought they both owned a VM with ID = XYZ.

Do you think this could be related?

**Comment by Tinyblargon** [comment]
@zapotocnylubos it could be related, but rc6 sets the amount of threads to `1` by default.

**Comment by Tinyblargon** [comment]
@zapotocnylubos running multiple terraforms in parralel is always going to be an issue. As we can only have a lock in terraform or the client session. In your specific use case, I'd look into assigning a specific id to each vm.

**Comment by TomyLobo** [comment]
> @zapotocnylubos running multiple terraforms in parralel is always going to be an issue. As we can only have a lock in terraform or the client session. In your specific use case, I'd look into assigning a specific id to each vm.

It should not be an issue, though, as long as you don't manage the _same_ resource in two parallel terraform runs.

**Comment by zapotocnylubos** [comment]
I think I see the problem - it was unclear to me why do we need to call `nextid` before `POST`ing  and creating the VM.

As @TomyLobo showed in

> Here's what I got using an MITM proxy:
> ...

By looking at the [Proxmox docs](https://pve.proxmox.com/pve-docs/api-viewer/#/nodes/{node}/qemu), the API for creation of VM requires a `vmid` as a parameter, there is no "autogeneration" option right?

<img width="806" alt="Snímek obrazovky 2024-12-06 v 12 25 53" src="https://github.com/user-attachments/assets/101f4d4d-a699-42c7-b536-ead05c725bd9">

That's why @Tinyblargon suggested an "external" way of dealing with assigning IDs in parallel deployment scenarios. Right?

**Comment by Tinyblargon** [comment]
@zapotocnylubos Yes, when running parralel deployments (multiple parralel terraform apply), you should handle the ID. When running, just one terraform apply over multiple threads the provider should be able to handle it (due to an implementation issue it can't).

To fix the treading for 1 terraform apply, some changes have to be made in `proxmox-api-go` and in this project (I'm already working on those).

It's also possible to deal with the multiple parralel providers in `proxmox-api-go`, but that would be a lot of work and isn't a blocker for the official 3.0 release.

**Comment by TomyLobo** [comment]
Ideally all VMID assignment would be handled by Proxmox itself.

POST /api2/json/nodes/{node}/qemu needs to make the vmid optional and then fetch one from whatever source GET /api2/json/cluster/nextid uses.
The section from fetching an ID to occupying it with a VM needs to be mutexed on proxmox's side, so multiple concurrent requests can't steal each others' IDs

Btw, @zapotocnylubos are you zapotah on ##proxmox? :)

**Comment by zapotocnylubos** [comment]
@TomyLobo I am not.

I also expected that the auto-generation of IDs using Proxmox API would be parallel-safe (generated and published as an API call response if not provided as a parameter), but it's not apparently.

**Comment by Tinyblargon** [comment]
@lotgyero creation of LXC still gives conflicts. 
When you only create Qemu instances, it should be safe, with the current master branch.

**Comment by Tinyblargon** [comment]
@zapotocnylubos could you test if the ID conflicts still occur with 3.0.1rc7? The creation of Qemu guests should be able to deal with multiple providers in parallel.

**Comment by zapotocnylubos** [comment]
@Tinyblargon do you mean, that when I run (in my case) 4+ parallel jobs (isolated runs) of `terraform apply` for different projects (all connected to the same Proxmox though) I should see consistent vmid provisioning and no conflicts?

**Comment by Tinyblargon** [comment]
@zapotocnylubos exactly, but only when they exclusively provision Qemu guests, LXC guests still have treading issues.

**Comment by Tinyblargon** [comment]
For create Lxc and Qemu we have integration tests for this race condition upstream.

https://github.com/Telmate/proxmox-api-go/blob/9cba66824699e24334db714e69237cf2a3b04d2e/test/api/lxc/create_lxc_test.go#L155
https://github.com/Telmate/proxmox-api-go/blob/9cba66824699e24334db714e69237cf2a3b04d2e/test/api/qemu/create_qemu_test.go#L20

We still have to add those for Lxc and Qemu clone.
